### PR TITLE
Laravel 5 Compatibility

### PR DESCRIPTION
**Pull request by tomschlick**
The existing L4 stuff should probably go into a laravel4 branch to maintain them separately.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,21 +27,29 @@ php composer.phar update
 
 To use the AWS Service Provider, you must register the provider when bootstrapping your Laravel application.
 
-Publish the package configuration using Artisan.
+By default, the package uses the following environment variables to auto-configure the plugin without modification:
+```
+AWS_KEY
```
Comment (jeremeamia) [inline README.md]: We use `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY ` in the SDK, so it would make sense to standardize on that, right?

**Comment by tomschlick** [comment]
Yeah thats a good point. I'll make the changes.

File: composer.json [hunk 2]
```diff
@@ -14,13 +14,12 @@
...
-        "illuminate/support": "4.*"
+        "illuminate/support": "~5.0"
     },
     "require-dev": {
-        "phpunit/phpunit": "3.7.*"
+        "phpunit/phpunit": "~4.0"
     },
     "autoload": {
-        "psr-0": { "Aws\\Laravel": "src/" }
+        "psr-4": { "Aws\\Laravel\\": "src/Aws/Laravel/" }
```
Comment (GrahamCampbell) [inline composer.json]: We can probably move the files up the dirs too.

File: src/Aws/Laravel/AwsServiceProvider.php [hunk 3]
```diff
@@ -37,30 +44,41 @@ class AwsServiceProvider extends ServiceProvider
...
     }
 
+    public function prepareConfig()
+    {
+        $config = realpath(__DIR__ . '/../../config/config.php');
+
+        $this->mergeConfigFrom($config, 'aws');
+
+        $this->publishes([
+                             $config => config_path('aws.php'),
```
Comment (GrahamCampbell) [inline src/Aws/Laravel/AwsServiceProvider.php]: Indentation totally wrong here.

File: src/Aws/Laravel/AwsServiceProvider.php [hunk 4]
```diff
@@ -37,30 +44,41 @@ class AwsServiceProvider extends ServiceProvider
      */
     public function register()
     {
-        $this->app['aws'] = $this->app->share(function ($app) {
-            // Retrieve the config
-            $config = $app['config']['aws'] ?: $app['config']['aws::config'];
-            if (isset($config['config_file'])) {
-                $config = $config['config_file'];
-            }
+        $this->prepareConfig();
```
Comment (GrahamCampbell) [inline src/Aws/Laravel/AwsServiceProvider.php]: This needs to be in the boot method, not register,

File: src/config/config.php [hunk 5]
```diff
@@ -30,8 +30,8 @@
     | http://docs.aws.amazon.com/aws-sdk-php-2/guide/latest/configuration.html
     |
     */
-    'key'    => null, // Your AWS Access Key ID
-    'secret' => null, // Your AWS Secret Access Key
+    'key'         => env('AWS_ACCESS_KEY_ID'),
```
Comment (GrahamCampbell) [inline src/config/config.php]: Too many spaces!

File: src/AwsServiceProvider.php [hunk 6]
```diff
@@ -37,30 +44,39 @@ class AwsServiceProvider extends ServiceProvider
...
 
             // Attach an event listener that will append the Laravel and module version numbers to the user agent string
-            $aws->getEventDispatcher()->addListener('service_builder.create_client', function (Event $event) {
-                $clientConfig = $event['client']->getConfig();
-                $commandParams = $clientConfig->get(Client::COMMAND_PARAMS) ?: array();
-                $userAgentSuffix = 'Laravel/' . Application::VERSION . ' L4MOD/' . AwsServiceProvider::VERSION;
-                $clientConfig->set(Client::COMMAND_PARAMS, array_merge_recursive($commandParams, array(
+            $aws->getEventDispatcher()->addListener('service_builder.create_client', function (Event $event)
+            {
+                $clientConfig       = $event['client']->getConfig();
```
Comment (jeremeamia) [inline src/AwsServiceProvider.php]: This indentation seems weird.

File: src/AwsServiceProvider.php [hunk 7]
```diff
@@ -37,30 +44,39 @@ class AwsServiceProvider extends ServiceProvider
      */
     public function register()
     {
-        $this->app['aws'] = $this->app->share(function ($app) {
-            // Retrieve the config
-            $config = $app['config']['aws'] ?: $app['config']['aws::config'];
-            if (isset($config['config_file'])) {
-                $config = $config['config_file'];
-            }
+        $this->app['aws'] = $this->app->share(function ($app)
```
Comment (jeremeamia) [inline src/AwsServiceProvider.php]: Shouldn't we use `$this->app->singleton('aws', function($app) {...});` like in the [docs](http://laravel.com/docs/master/providers#basic-provider-example)?

File: src/AwsServiceProvider.php (hunk 6 above)
Comment (tomschlick) [inline src/AwsServiceProvider.php]: Sorry, used Sublime Text for the first time in a while and its seemed to cause all sorts of indentation issues. I'll fix it in a few minutes.

**Comment by tomschlick** [comment]
Anything else needed to get this merged?

**Comment by jeremeamia** [comment]
@tomschlick Do we need to incorporate #51 in anyway?

Also, can we squash the commits for this PR?

@taylorotwell what do you think about the changes here?

File: src/AwsServiceProvider.php [hunk 8]
```diff
@@ -26,9 +26,16 @@
 /**
  * AWS SDK for PHP service provider for Laravel applications
  */
-class AwsServiceProvider extends ServiceProvider
-{
-    const VERSION = '1.1.0';
+class AwsServiceProvider extends ServiceProvider {
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: Braces should be on a new line for classes. PSR-2.

File: src/AwsServiceProvider.php [hunk 9]
```diff
@@ -37,30 +44,37 @@ class AwsServiceProvider extends ServiceProvider
...
-            $aws = Aws::factory($config);
+            $aws = Aws::factory($app['config']->get('aws'));
 
             // Attach an event listener that will append the Laravel and module version numbers to the user agent string
             $aws->getEventDispatcher()->addListener('service_builder.create_client', function (Event $event) {
-                $clientConfig = $event['client']->getConfig();
-                $commandParams = $clientConfig->get(Client::COMMAND_PARAMS) ?: array();
-                $userAgentSuffix = 'Laravel/' . Application::VERSION . ' L4MOD/' . AwsServiceProvider::VERSION;
-                $clientConfig->set(Client::COMMAND_PARAMS, array_merge_recursive($commandParams, array(
+                $clientConfig       = $event['client']->getConfig();
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: These shouldn't be aligned.

File: src/AwsServiceProvider.php [hunk 10]
```diff
@@ -37,30 +44,37 @@ class AwsServiceProvider extends ServiceProvider
...
                     UserAgentListener::OPTION => $userAgentSuffix,
-                )));
+                ]));
             });
 
             return $aws;
         });
     }
 
+    public function prepareConfig()
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: This doesn't need it's own method. Just put it in the boot method. It's also convention to have the boot method above the register method.

File: src/AwsServiceProvider.php [hunk 11]
```diff
@@ -37,30 +44,37 @@ class AwsServiceProvider extends ServiceProvider
...
+
+        $this->mergeConfigFrom($config, 'aws');
+
+        $this->publishes([
+             $config => config_path('aws.php'),
+             ], 'config');
+    }
+
     /**
      * Bootstrap the application events.
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: This description is wrong.

**Comment by tomschlick** [comment]
ok this should be good to go from the latest suggestions

File: src/AwsServiceProvider.php [hunk 12]
```diff
@@ -28,7 +28,31 @@
...
+     *
+     * @return void
+     */
+    public function boot()
+    {
+        $config = realpath(__DIR__ . '/../config/config.php');
+
+        $this->mergeConfigFrom($config, 'aws');
+
+        $this->publishes([
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: This is die one one line.

Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: This is fine on one line.

File: src/AwsServiceProvider.php [hunk 13]
```diff
@@ -37,40 +61,26 @@ class AwsServiceProvider extends ServiceProvider
      */
     public function register()
     {
-        $this->app['aws'] = $this->app->share(function ($app) {
-            // Retrieve the config
-            $config = $app['config']['aws'] ?: $app['config']['aws::config'];
-            if (isset($config['config_file'])) {
-                $config = $config['config_file'];
-            }
+        $this->app->singleton('aws', function ($app) {
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: This extra blank line needs removing.

File: src/AwsServiceProvider.php [hunk 14]
```diff
@@ -80,6 +90,6 @@ public function boot()
      */
     public function provides()
     {
-        return array('aws');
+        return ['aws'];
```
Comment (GrahamCampbell) [inline src/AwsServiceProvider.php]: You need to add `'Aws\Common\Aws'` to this array too.

File: composer.json [hunk 15]
```diff
@@ -14,13 +14,12 @@
     "require": {
         "php": ">=5.3.3",
```
Comment (GrahamCampbell) [inline composer.json]: Should be `">=5.4.0"` now that we're using short array syntax in here.

**Comment by tomschlick** [comment]
Anything additional holding this up?

**Comment by jeremeamia** [comment]
Well, the tests don't work. :confused: I'm going to take a look at this today.

Also, I saw a [competing L5 AWS provider](https://github.com/caseyh/aws-sdk-php-laravel-5). @caseyh, have you seen what we are doing here? Want to merge projects?

**Comment by tomschlick** [comment]
Sounds good. Adding Travis CI would definitely help keep those in check too.

**Comment by caseyh** [comment]
Hi! Sorry, I was just very frustrated with the whole setup (I'm a very recent codeigniter convert) so I decided to do my own thing until something better that worked arrived. I hadn't see this though. I'd be happy to delete my packagist for it, assuming this merge is happening in a reasonable amount of time. I'm pretty sure my version does nothing additional or better than yours.
The only thing is that I tried to start working on tests but being new to Laravel, I haven't found the answer to the getConfigLoader() issue yet.

**Comment by jeremeamia** [comment]
I got the tests figured out. I need to create a branch for 1.x. Then I'll merge this in and apply my changes.

**Comment by caseyh** [comment]
Bless!

**Comment by jeremeamia** [comment]
@tomschlick I just PR'd this PR: https://github.com/runmybusiness/aws-sdk-php-laravel/pull/1

**Comment by tomschlick** [comment]
Merged!

**Comment by jeremeamia** [comment]
Merged!

**Comment by jeremeamia** [comment]
Give the [AWS Provider for L5 (Beta)](https://github.com/aws/aws-sdk-php-laravel/releases/tag/2.0.0-beta.1) a try.

**Pull request merged**